src/cmd/services/m3coordinator/downsample/downsampler.go

```diff
@@ -0,0 +1,757 @@
+		encodedTagsIteratorPool: encodedTagsIteratorPool,
+		workerPool:              workerPool,
+		instrumentOpts:          instrumentOpts,
+		metrics:                 newDownsamplerFlushHandlerMetrics(scope),
+	}
+}
+
+func (h *downsamplerFlushHandler) NewWriter(
+	scope tally.Scope,
+) (writer.Writer, error) {
+	return &downsamplerFlushHandlerWriter{
+		ctx:     context.Background(),
+		handler: h,
+	}, nil
+}
+
+func (h *downsamplerFlushHandler) Close() {
+}
+
+type downsamplerFlushHandlerWriter struct {
+	wg      sync.WaitGroup
+	ctx     context.Context
+	handler *downsamplerFlushHandler
+}
+
+func (w *downsamplerFlushHandlerWriter) Write(
+	mp aggregated.ChunkedMetricWithStoragePolicy,
+) error {
+	w.wg.Add(1)
+	w.handler.workerPool.Go(func() {
+		defer w.wg.Done()
+
+		logger := w.handler.instrumentOpts.Logger()
+
+		iter := w.handler.encodedTagsIteratorPool.Get()
+		iter.Reset(mp.ChunkedID.Data)
+		expected := iter.tagDecoder.Remaining()
+		if len(mp.ChunkedID.Suffix) > 0 {
+			expected++
+		}
+		tags := make(models.Tags, expected)
+		for iter.Next() {
+			name, value := iter.Current()
+			tags[string(name)] = string(value)
+		}
+		if len(mp.ChunkedID.Suffix) > 0 {
+			tags[aggregationSuffixTag] = string(mp.ChunkedID.Suffix)
+		}
+
+		err := iter.Err()
+		iter.Close()
+		if err != nil {
+			logger.Debugf("downsampler flush error preparing write: %v", err)
+			w.handler.metrics.flushErrors.Inc(1)
+			return
+		}
+
+		err = w.handler.storage.Write(w.ctx, &storage.WriteQuery{
+			Tags: tags,
+			Datapoints: ts.Datapoints{ts.Datapoint{
+				Timestamp: time.Unix(0, mp.TimeNanos),
+				Value:     mp.Value,
+			}},
+			Unit: mp.StoragePolicy.Resolution().Precision,
+			Attributes: storage.Attributes{
+				MetricsType: storage.AggregatedMetricsType,
+				Retention:   mp.StoragePolicy.Retention().Duration(),
+				Resolution:  mp.StoragePolicy.Resolution().Window,
+			},
+		})
+		if err != nil {
+			logger.Debugf("downsampler flush error failed write: %v", err)
+			w.handler.metrics.flushErrors.Inc(1)
+			return
+		}
+
+		w.handler.metrics.flushSuccess.Inc(1)
+	})
+
+	return nil
+}
+
+func (w *downsamplerFlushHandlerWriter) Flush() error {
+	w.wg.Wait()
+	return nil
+}
+
+func (w *downsamplerFlushHandlerWriter) Close() error {
+	return nil
+}
+
+type downsampler struct {
+	cfg        DownsamplingConfiguration
+	aggregator newAggregatorResult
+}
+
+// NewDownsampler returns a new downsampler.
+func NewDownsampler(
+	cfg DownsamplingConfiguration,
+	opts DownsamplerOptions,
+) (Downsampler, error) {
+	d := &downsampler{
+		cfg: cfg,
+	}
+
+	if cfg.AggregationPolicy.Enabled {
+		result, err := opts.newAggregator()
+		if err != nil {
+			return nil, err
+		}
+		d.aggregator = result
+	}
+
+	return d, nil
+}
+
+func (d *downsampler) NonePolicy() NonePolicy {
+	return d.cfg.NonePolicy
+}
+
+func (d *downsampler) AggregationPolicy() AggregationPolicy {
+	return d.cfg.AggregationPolicy
+}
+
+func (d *downsampler) MetricsAppender() MetricsAppender {
```

Comment:
nit: I think this would make more sense as `NewMetricsAppender()`, would also make it clearer that you're getting a copy of the appender inside the downsampler goroutines